File: pathfinding/a-star/visual-basic/a_star.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


Module program

    ' Index values for g-score, f-score and previous node
    Public Const GSCORE As Integer = 0
    Public Const FSCORE As Integer = 1
    Public Const PREVIOUS As Integer = 2


    ' The Main method is the default entry point for all VB programs
    Sub Main()
        ' Use a dictionary to represent the graph as an adjacency list
        ' and the g-score and f-score of each neighbour
        Dim testGraph = New Dictionary(Of String, Dictionary(Of String, Integer)) From {
            {"A", New Dictionary(Of String, Integer) From {{"B", 10}, {"C", 12}, {"D", 5}}},
            {"B", New Dictionary(Of String, Integer) From {{"A", 10}, {"E", 11}}},
            {"C", New Dictionary(Of String, Integer) From {{"A", 12}, {"D", 6}, {"E", 11}, {"F", 8}}},
            {"D", New Dictionary(Of String, Integer) From {{"A", 5}, {"C", 6}, {"F", 14}}},
            {"E", New Dictionary(Of String, Integer) From {{"B", 11}, {"C", 11}}},
            {"F", New Dictionary(Of String, Integer) From {{"C", 8}, {"D", 14}}}
        }

        Console.WriteLine("### A* algorithm ###")
        Console.WriteLine(vbLf & "Display graph: " & vbLf)
        DisplayGraph(testGraph)

        ' Perform the A* algorithm between nodes A and F
        Dim visitedList = AStar(testGraph, "A", "F")

        ' Display the shortest path using the visited list
        DisplayShortestPath(visitedList, "A", "F")
    End Sub


    ' Display each node with it's neighbours and costs
    Sub DisplayGraph(ByVal graph As Dictionary(Of String, Dictionary(Of String, Integer)))
        ' Repeat for each node in the graph
        For Each kvp In graph
            Dim node = kvp.Key
            Dim neighbours = kvp.Value

            Console.WriteLine($"Node: {kvp.Key}")
            Console.Write("Neighbours: ")

            ' Repeat for each neighbour node in the neighbours list
            For Each n_node In neighbours
                Console.Write($"{n_node.Key}:{n_node.Value} ")
            Next
            Console.WriteLine(vbLf)
        Next
    End Sub


    ' Display a list of nodes with their closest neighbour and scores
    Sub DisplayList(ByVal adjacencyList As Dictionary(Of String, List(Of Object)))
        Console.WriteLine("   (g-score, f-score, previous)")

        ' Repeat for each node in the given adjacency list
        For Each kvp In adjacencyList
            Dim node = kvp.Key
            Dim neighbours = kvp.Value

            ' If the neighbour node (nNode) does not exist then set the string to "Nothing"
            Dim nNode = If(neighbours(PREVIOUS) Is Nothing, Nothing, CStr(neighbours(PREVIOUS)))

            Console.WriteLine($"{node}: {neighbours(GSCORE)}, {neighbours(FSCORE)}, {nNode}")
        Next
        Console.WriteLine()
    End Sub

    ' Display the shortest path from start node to target node
    Sub DisplayShortestPath(ByVal visited As Dictionary(Of String, List(Of Object)), ByVal startNode As String, ByVal targetNode As String)
        ' Set the current node and the path as the key node
        Dim currentNode = targetNode
        Dim path = targetNode

        ' Repeat until the current node reaches the start node
        While Not Equals(currentNode, startNode)
            ' Add the previous node to the start of the path
            Dim previousNode = CStr(visited(currentNode)(PREVIOUS))
            path = previousNode & path

            ' Update the current node to be the previous node
            currentNode = previousNode
        End While

        ' Testing
        Dim cost As Integer = visited(targetNode)(GSCORE)
        Console.WriteLine($"The shortest path from {startNode} to {targetNode} is:")
        Console.WriteLine($"Path: {path}")
        Console.WriteLine($"Cost: {cost}")
    End Sub

    ' Returns heuristic values for the graph as used on the Isaac CS platform
    Function GetHeuristic(ByVal node As String) As Integer
        Dim estimatedDistanceToTarget As Integer

        If Equals(node, "A") Then
            estimatedDistanceToTarget = 10
        ElseIf Equals(node, "B") Then
            estimatedDistanceToTarget = 15
        ElseIf Equals(node, "C") Then
            estimatedDistanceToTarget = 5
        ElseIf Equals(node, "D") Then
            estimatedDistanceToTarget = 5
        ElseIf Equals(node, "E") Then
            estimatedDistanceToTarget = 10
        ElseIf Equals(node, "F") Then
            estimatedDistanceToTarget = 0
        Else
            estimatedDistanceToTarget = 0
        End If

        Return estimatedDistanceToTarget
    End Function

    ' Returns the node with the lowest f-score
    Function GetMinimum(ByVal unvisited As Dictionary(Of String, List(Of Object))) As String
        ' Set the lowest value to Int32.MaxValue for infinity
        Dim lowestValueNode = Integer.MaxValue
        Dim lowestKey = Nothing

        ' Repeat for each node in the unvisited list
        For Each kvp In unvisited
            Dim node = kvp.Value

            ' Get the f-score value
            Dim fScoreValue As Integer = node(FSCORE)

            ' Check if the f-score is less than the lowest value
            If fScoreValue < lowestValueNode Then
                ' Update lowest value and lowest key
                lowestValueNode = fScoreValue
                lowestKey = kvp.Key
            End If
        Next
        Return lowestKey
    End Function

    ' Apply the A* algorithm on a graph stored as a dictionary
    Function AStar(ByVal graph As Dictionary(Of String, Dictionary(Of String, Integer)), ByVal startNode As String, ByVal targetNode As String) As Dictionary(Of String, List(Of Object))
        ' Declare the visited and unvisited lists as dictionaries
        Dim unvisited As Dictionary(Of String, List(Of Object)) = New Dictionary(Of String, List(Of Object))()
        Dim visited As Dictionary(Of String, List(Of Object)) = New Dictionary(Of String, List(Of Object))()

        ' Add and initialise every node to the unvisited list
        For Each kvp In graph
            ' Get the node
            Dim node = kvp.Key

            ' Initialise g-score and f-score to Int32.MaxValue (for infinity)
            '  and the previous node to null
            unvisited(node) = New List(Of Object)() From {Integer.MaxValue, Integer.MaxValue, Nothing}
        Next

        ' Update the values for the start node in the unvisited list
        Dim fScoreValue = GetHeuristic(startNode)
        unvisited(startNode) = New List(Of Object)() From {0, fScoreValue, Nothing}

        ' Testing
        Console.WriteLine("--- Initialised state of unvisited list ---")
        DisplayList(unvisited)

        ' Repeat until there are no more nodes in the unvisited list
        Dim finished = False
        While finished = False
            ' Check if there are no more nodes left to evaluate
            If unvisited.Count = 0 Then
                finished = True
            Else
                ' Return the unvisited node with the lowest f-score
                Dim currentNode = GetMinimum(unvisited)
                Console.WriteLine($"Current node >>> {currentNode}") ' Testing

                ' Check if the current node is the target node
                If Equals(currentNode, targetNode) Then
                    ' Add the current node to the visited list
                    finished = True
                    visited(currentNode) = unvisited(currentNode)
                Else
                    ' Get the current node's list of neighbours
                    Dim neighbours = graph(currentNode)

                    ' Repeat for each neighbour node in the neighbours list
                    For Each kvp In neighbours
                        ' Get the neighbour node
                        Dim node = kvp.Key

                        ' Check if the neighbour node has already been visited
                        If visited.ContainsKey(node) = False Then
                            ' Calculate the new g-score
                            Dim newGScore = CInt(unvisited(currentNode)(GSCORE)) + neighbours(node)

                            ' Check if the new g-score is less
                            If newGScore < CInt(unvisited(node)(GSCORE)) Then
                                ' Update g-score, f-score and previous node
                                unvisited(node)(GSCORE) = newGScore
                                unvisited(node)(FSCORE) = newGScore + GetHeuristic(node)
                                unvisited(node)(PREVIOUS) = currentNode
                            End If
                        End If
                    Next
                    ' Testing
                    Console.WriteLine("--- Unvisited list ---")
                    DisplayList(unvisited)

                    ' Add the current node to the visited list
                    visited(currentNode) = unvisited(currentNode)

                    ' Remove the current node from the unvisited list
                    unvisited.Remove(currentNode)

                    ' Testing
                    Console.WriteLine($"Moved {currentNode} to visited list")
                    Console.WriteLine("--- Visited list ---")
                    DisplayList(visited)
                End If
            End If
        End While
        ' Testing
        Console.WriteLine("--- Visited list ---")
        DisplayList(visited)

        ' Return the final visited list
        Return visited
    End Function


End Module

File: pathfinding/a-star/java/a_star.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Pathfinding {
    // Index values for g-score, f-score and previous node
    public static final int G_SCORE = 0;
    public static final int F_SCORE = 1;
    public static final int PREVIOUS = 2;

    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Use a hashmap to represent the graph as an adjacency list
        // and the g-score and f-score of each neighbour
        var testGraph = new HashMap<String, HashMap<String, Integer>>();

        var nodeA = new HashMap<String, Integer>();
        nodeA.put("B", 10);
        nodeA.put("C", 12);
        nodeA.put("D", 5);
        testGraph.put("A", nodeA);

        var nodeB = new HashMap<String, Integer>();
        nodeB.put("A", 10);
        nodeB.put("E", 11);
        testGraph.put("B", nodeB);

        var nodeC = new HashMap<String, Integer>();
        nodeC.put("A", 12);
        nodeC.put("D", 6);
        nodeC.put("E", 11);
        nodeC.put("F", 8);
        testGraph.put("C", nodeC);

        var nodeD = new HashMap<String, Integer>();
        nodeD.put("A", 5);
        nodeD.put("C", 6);
        nodeD.put("F", 14);
        testGraph.put("D", nodeD);

        var nodeE = new HashMap<String, Integer>();
        nodeE.put("B", 11);
        nodeE.put("C", 11);
        testGraph.put("E", nodeE);

        var nodeF = new HashMap<String, Integer>();
        nodeF.put("C", 8);
        nodeF.put("D", 14);
        testGraph.put("F", nodeF);

        System.out.println("### A* algorithm ###");
        System.out.println("\nDisplay graph: \n");
        displayGraph(testGraph);

        // Perform the A* algorithm between nodes A and F
        var visitedList = aStar(testGraph, "A", "F");

        // Display the shortest path using the visited list
        displayShortestPath(visitedList, "A", "F");
    }

    // Display each node with it's neighbours and costs
    public static void displayGraph(HashMap<String, HashMap<String, Integer>> graph) {
        // Repeat for each node in the graph
        for (String node : graph.keySet()) {
            HashMap<String, Integer> neighbours = graph.get(node);

            System.out.println("Node: " + node);
            System.out.print("Neighbours: ");

            // Repeat for each neighbour node in the neighbours list
            for (String n_node : neighbours.keySet()) {
                System.out.print(n_node + ":" + neighbours.get(n_node) + " ");
            }
            System.out.println("\n");
        }
    }

    // Display a list of nodes with their closest neighbour and scores
    public static void displayList(HashMap<String, ArrayList<Object>> adjacencyList) {
        System.out.println("   (g-score, f-score, previous)");

        // Repeat for each node in the given adjacency list
        for (String node : adjacencyList.keySet()) {
            ArrayList<Object> neighbours = adjacencyList.get(node);

            // If the neighbour node (nNode) does not exist then set the String to "null"
            String nNode = neighbours.get(PREVIOUS) == null ? "null" : (String)neighbours.get(PREVIOUS);

            System.out.println(node + ": " + neighbours.get(G_SCORE) + ", " + neighbours.get(F_SCORE) + ", " + nNode);
        }
        System.out.println();
    }

    // Display the shortest path from start node to target node
    public static void displayShortestPath(HashMap<String, ArrayList<Object>> visited,
                                           String startNode, String targetNode) {
        // Set the current node and the path as the key node
        String currentNode = targetNode;
        String path = targetNode;

        // Repeat until the current node reaches the start node
        while (!currentNode.equals(startNode)) {
            // Add the previous node to the start of the path
            String previousNode = (String)visited.get(currentNode).get(PREVIOUS);
            path = previousNode + path;

            // Update the current node to be the previous node
            currentNode = previousNode;
        }

        // Testing
        int cost = (int)visited.get(targetNode).get(G_SCORE);
        System.out.println("The shortest path from " + startNode + " to " + targetNode + " is:");
        System.out.println("Path: " + path);
        System.out.println("Cost: " + cost);
    }

    // Returns heuristic values for the graph as used on the Ada CS platform
    public static int getHeuristic(String node) {
        int estimatedDistanceToTarget;

        if ("A".equals(node)) {
            estimatedDistanceToTarget = 10;
        }
        else if ("B".equals(node)) {
            estimatedDistanceToTarget = 15;
        }
        else if ("C".equals(node)) {
            estimatedDistanceToTarget = 5;
        }
        else if ("D".equals(node)) {
            estimatedDistanceToTarget = 5;
        }
        else if ("E".equals(node)) {
            estimatedDistanceToTarget = 10;
        }
        else if ("F".equals(node)) {
            estimatedDistanceToTarget = 0;
        }
        else {
            estimatedDistanceToTarget = 0;
        }

        return estimatedDistanceToTarget;
    }

    // Returns the node with the lowest f-score
    public static String getMinimum(HashMap<String, ArrayList<Object>> unvisited) {
        // Set the lowest value to Integer.MAX_VALUE for infinity
        int lowestValueNode = Integer.MAX_VALUE;
        String lowestKey = "";

        // Repeat for each node in the unvisited list
        for (Map.Entry<String, ArrayList<Object>> entry : unvisited.entrySet()) {
            ArrayList<Object> node = entry.getValue();

            // Get the f-score value
            int fScoreValue = (int)node.get(F_SCORE);

            // Check if the f-score is less than the lowest value
            if (fScoreValue < lowestValueNode) {
                // Update lowest value and lowest key
                lowestValueNode = fScoreValue;
                lowestKey = entry.getKey();
            }
        }
        return lowestKey;
    }

    // Apply the A* algorithm on a graph stored as a hashmap
    public static HashMap<String, ArrayList<Object>> aStar(
            HashMap<String, HashMap<String, Integer>> graph, String startNode, String targetNode) {
        // Declare the visited and unvisited lists as hashmaps
        HashMap<String, ArrayList<Object>> unvisited = new HashMap<String, ArrayList<Object>>();
        HashMap<String, ArrayList<Object>> visited = new HashMap<String, ArrayList<Object>>();

        // Add and initialise every node to the unvisited list
        for (String node : graph.keySet()) {
            // Initialise g-score and f-score to Integer.MAX_VALUE (for infinity)
            // and the previous node to null
            unvisited.put(node, new ArrayList<Object>(Arrays.asList(Integer.MAX_VALUE, Integer.MAX_VALUE, null)));
        }

        // Update the values for the start node in the unvisited list
        int fScoreValue = getHeuristic(startNode);
        unvisited.put(startNode, new ArrayList<Object>(Arrays.asList(0, fScoreValue, null)));

        // Testing
        System.out.println("--- Initialised state of unvisited list ---");
        displayList(unvisited);

        // Repeat until there are no more nodes in the unvisited list
        boolean finished = false;
        while (finished == false) {
            // Check if there are no more nodes left to evaluate
            if (unvisited.size() == 0)
                finished = true;
            else {
                // Return the unvisited node with the lowest f-score
                String currentNode = getMinimum(unvisited);
                System.out.println("\nCurrent node >>> " + currentNode); // Testing

                // Check if the current node is the target node
                if (currentNode.equals(targetNode)) {
                    // Add the current node to the visited list
                    finished = true;
                    visited.put(currentNode, unvisited.get(currentNode));
                }
                else {
                    // Get the current node's list of neighbours
                    HashMap<String, Integer> neighbours = graph.get(currentNode);

                    // Repeat for each neighbour node in the neighbours list
                    for (String node : neighbours.keySet()) {
                        // Check if the neighbour node has already been visited
                        if (visited.containsKey(node) == false) {
                            // Calculate the new g-score
                            int newGScore = (int)unvisited.get(currentNode).get(G_SCORE) + neighbours.get(node);

                            // Check if the new g-score is less
                            if (newGScore < (int)unvisited.get(node).get(G_SCORE)) {
                                // Update g-score, f-score and previous node
                                unvisited.get(node).set(G_SCORE, newGScore);
                                unvisited.get(node).set(F_SCORE, newGScore + getHeuristic(node));
                                unvisited.get(node).set(PREVIOUS, currentNode);
                            }
                        }
                    }
                    // Testing
                    System.out.println("--- Unvisited list ---");
                    displayList(unvisited);

                    // Add the current node to the visited list
                    visited.put(currentNode, unvisited.get(currentNode));

                    // Remove the current node from the unvisited list
                    unvisited.remove(currentNode);

                    // Testing
                    System.out.println("Moved " + currentNode + " to visited list");
                    System.out.println("--- Visited list ---");
                    displayList(visited);
                }
            }
        }
        // Testing
        System.out.println("--- Visited list ---");
        displayList(visited);

        // Return the final visited list
        return visited;
    }


}